Auth: Add skip_org_role_sync setting for github

eleijonmarck opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

this add the `skip_org_role_sync` for the Github OAuth provider

**Which issue(s) does this PR fix?**:
https://github.com/grafana/grafana-enterprise/issues/4235

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer**:

demo - 


https://user-images.githubusercontent.com/7727602/214007628-7c224b9e-accb-4602-ac19-d5a5b62e2e4b.mp4

eleijonmarck committed:
add: skip_org_role_sync setting for github

eleijonmarck committed:
fix: frontend

eleijonmarck committed:
fix merge

eleijonmarck committed:
Merge branch 'main' into eleijonmarck/auth/github-skip-org-role-sync

eleijonmarck committed:
Merge branch 'main' into eleijonmarck/auth/github-skip-org-role-sync

eleijonmarck committed:
rearranged tests

IevaVasiljeva commented on pkg/login/social/github_oauth.go:
```diff
@@ -201,9 +203,15 @@ func (s *SocialGithub) UserInfo(client *http.Client, token *oauth2.Token) (*Basi
-	if s.roleAttributeStrict && !role.IsValid() {
-		return nil, &InvalidBasicRoleError{idP: "Github", assignedRole: string(role)}
+	var role roletype.RoleType
+	var grafanaAdmin bool
```
I think currently you would always set Grafana server admin role to false if Github `skipOrgRoleSync` and `allowAssignGrafanaAdmin` are both enabled. It's a bit of a weird config setup, but I think we'd want to skip syncing server admin role in that case, right?

So you probably want to set `IsGrafanaAdmin` to `nil` if Github `skipOrgRoleSync` is enabled to avoid syncing that role. 

It would also be good to have tests covering server admin.

eleijonmarck replied:
Right, so even though we set server admin role to false we would want to set `IsGrafanaAdmin` to `nil`. 
https://github.com/grafana/grafana/blob/main/pkg/login/social/social.go#L237

Comments like these make my heart go ❤️ 

I'l add some tests covering this scenario. Should we also then set some debugging or maybe logs for stating that you have set both `allowAssignGrafanaAdmin` and `skipOrgRoleSync` where `skipOrgRoleSync` disables the assignment of GrafanaAdmin

eleijonmarck replied:
@IevaVasiljeva could you have another look? 🙏

IevaVasiljeva commented on pkg/login/social/github_oauth.go:
```diff
@@ -7,6 +7,7 @@ import (
 	"net/http"
 	"regexp"
 
+	"github.com/grafana/grafana/pkg/models/roletype"
```
nit: import ordering - move the external `"golang.org/x/oauth2"` import above the internal dependency and add an empty line.

eleijonmarck replied:
nice! thx

IevaVasiljeva reviewed (commented):
Overall looks good, but I've added a comment about sever admin syncing - I think the current behaviour is not what we'd want, but I'm also not 100% sure about it.

eleijonmarck committed:
refactor: assignGrafanaAdmin skip also

chri2547 commented on docs/sources/setup-grafana/configure-security/configure-authentication/github/index.md:
```diff
@@ -203,3 +203,16 @@ Your GitHub teams can be referenced in two ways:
+## Skip organization role sync
+
+To prevent the sync of org roles from Grafana.com, set `skip_org_role_sync` to `true`. This is useful if you want to manage the organization roles for your users from within Grafana.
+This also impacts `allow_assign_grafana_admin` setting, by not syncing the grafana admin role from GitHub.
```
```suggestion
This also impacts the `allow_assign_grafana_admin` setting by not syncing the Grafana admin role from GitHub.
```

chri2547 commented on docs/sources/setup-grafana/configure-security/configure-authentication/github/index.md:
```diff
@@ -203,3 +203,16 @@ Your GitHub teams can be referenced in two ways:
+
+## Skip organization role sync
+
+To prevent the sync of org roles from Grafana.com, set `skip_org_role_sync` to `true`. This is useful if you want to manage the organization roles for your users from within Grafana.
```
```suggestion
To prevent the sync of organization roles from Grafana.com, set `skip_org_role_sync` to `true`. This is useful if you want to manage the organization roles for your users from within Grafana.
```

chri2547 commented on docs/sources/setup-grafana/configure-grafana/_index.md:
```diff
@@ -904,6 +904,21 @@ The following table shows the OAuth provider's setting with the default value an
 
+### [auth.github] skip_org_role_sync
+
+Upon the first login from a user, we set the organization roles from the setting `AutoAssignOrgRole`. If you want to manage organizational roles, set the `skip_org_role_sync` option to `true`. Github syncs organizational roles and sets GrafanaAdmins as well.
```
```suggestion
When a user logs in the first time, Grafana sets the organization role based on the value specified in `AutoAssignOrgRole`. If you want to manage organization roles, set the `skip_org_role_sync` option to `true`. GitHub syncs organization roles and sets Grafana Admins.
```

chri2547 commented on docs/sources/setup-grafana/configure-grafana/_index.md:
```diff
@@ -904,6 +904,21 @@ The following table shows the OAuth provider's setting with the default value an
 
+### [auth.github] skip_org_role_sync
+
+Upon the first login from a user, we set the organization roles from the setting `AutoAssignOrgRole`. If you want to manage organizational roles, set the `skip_org_role_sync` option to `true`. Github syncs organizational roles and sets GrafanaAdmins as well.
```
what does "...sets Grafana Admins as well" mean?

eleijonmarck replied:
ah it means the role to be a GrafanaAdmin not just "Admin" as a user.

chri2547 commented on docs/sources/setup-grafana/configure-grafana/_index.md:
```diff
@@ -904,6 +904,21 @@ The following table shows the OAuth provider's setting with the default value an
+The following table shows the OAuth provider's setting with the default value and the skip org role sync setting.
+| OAuth Provider | `oauth_skip_org_role_sync_update` | `skip_org_role_sync` | Behavior |
+| --- | --- | --- | --- |
+| Github | false | false | User organization roles are set with `defaultRole` and cannot be changed |
```
```suggestion
| GitHub | false | false | User organization roles are set with `defaultRole` and cannot be changed |
```

chri2547 commented on docs/sources/setup-grafana/configure-grafana/_index.md:
```diff
@@ -904,6 +904,21 @@ The following table shows the OAuth provider's setting with the default value an
+| OAuth Provider | `oauth_skip_org_role_sync_update` | `skip_org_role_sync` | Behavior |
+| --- | --- | --- | --- |
+| Github | false | false | User organization roles are set with `defaultRole` and cannot be changed |
+| Github | true | false | User organization roles are set with `defaultRole` for Github, and GrafanaAdmins are set. For other providers, the synchronization will be skipped, and the org role can be changed, along with other OAuth provider users' org roles. |
```
```suggestion
| Github | true | false | User organization roles are set with `defaultRole` for GitHub, and Grafana Admins are set. For other providers, the synchronization is skipped, and the org role can be changed, along with other OAuth provider users' org roles. |
```

chri2547 commented on docs/sources/setup-grafana/configure-grafana/_index.md:
```diff
@@ -904,6 +904,21 @@ The following table shows the OAuth provider's setting with the default value an
+| --- | --- | --- | --- |
+| Github | false | false | User organization roles are set with `defaultRole` and cannot be changed |
+| Github | true | false | User organization roles are set with `defaultRole` for Github, and GrafanaAdmins are set. For other providers, the synchronization will be skipped, and the org role can be changed, along with other OAuth provider users' org roles. |
+| Github | false | true | User organization roles are set with `defaultRole` and the org role can be changed for Github synced users. |
```
```suggestion
| GitHub | false | true | User organization roles are set with `defaultRole`, and the organization role can be changed for GitHub synced users. |
```

chri2547 commented on docs/sources/setup-grafana/configure-grafana/_index.md:
```diff
@@ -904,6 +904,21 @@ The following table shows the OAuth provider's setting with the default value an
+| Github | false | false | User organization roles are set with `defaultRole` and cannot be changed |
+| Github | true | false | User organization roles are set with `defaultRole` for Github, and GrafanaAdmins are set. For other providers, the synchronization will be skipped, and the org role can be changed, along with other OAuth provider users' org roles. |
+| Github | false | true | User organization roles are set with `defaultRole` and the org role can be changed for Github synced users. |
+| Github | true | true | User organization roles are set with `defaultRole` for Google. For other providers, the synchronization will be skipped, and the org role can be changed, along with other OAuth provider users' org roles. |
```
```suggestion
| GitHub | true | true | User organization roles are set with `defaultRole` for Google. For other providers, the synchronization is skipped, and the org role can be changed, along with other OAuth provider users' org roles. |
```

chri2547 reviewed (commented):
Some suggestions and a question. Thank you!

mckn reviewed (commented):
LGTM!

eleijonmarck committed:
Add: tests for allowGrafanaAdmin

- both for the case when both settings are set and the setting for only
  allowGrafanaAdmin

eleijonmarck committed:
Apply suggestions from code review

Co-authored-by: Christopher Moyer <35463610+chri2547@users.noreply.github.com>

eleijonmarck committed:
Update docs/sources/setup-grafana/configure-grafana/_index.md

Co-authored-by: Christopher Moyer <35463610+chri2547@users.noreply.github.com>

IevaVasiljeva reviewed (approved):
Nice, LGTM! And thanks for adding the test case!

eleijonmarck committed:
Update pkg/login/social/github_oauth.go

Co-authored-by: Ieva <ieva.vasiljeva@grafana.com>

eleijonmarck committed:
added vairable inside scope

eleijonmarck commented on pkg/login/social/github_oauth.go:
```diff
@@ -201,14 +204,25 @@ func (s *SocialGithub) UserInfo(client *http.Client, token *oauth2.Token) (*Basi
-	if s.roleAttributeStrict && !role.IsValid() {
-		return nil, &InvalidBasicRoleError{idP: "Github", assignedRole: string(role)}
+	var role roletype.RoleType
+	var grafanaAdmin bool
```
i like it! makes it clear what variables we set where. nice one

eleijonmarck committed:
Merge branch 'main' into eleijonmarck/auth/github-skip-org-role-sync

eleijonmarck commented on docs/sources/setup-grafana/configure-security/configure-authentication/github/index.md:
````diff
@@ -203,3 +203,16 @@ Your GitHub teams can be referenced in two ways:
+This also impacts the `allow_assign_grafana_admin` setting by not syncing the Grafana admin role from GitHub.
+
+```ini
+[auth.grafana_com]
````
```suggestion
[auth.github]
```

eleijonmarck commented on docs/sources/setup-grafana/configure-security/configure-authentication/github/index.md:
```diff
@@ -203,3 +203,16 @@ Your GitHub teams can be referenced in two ways:
+
+## Skip organization role sync
+
+To prevent the sync of organization roles from Grafana.com, set `skip_org_role_sync` to `true`. This is useful if you want to manage the organization roles for your users from within Grafana.
```
```suggestion
To prevent the sync of organization roles from GitHub, set `skip_org_role_sync` to `true`. This is useful if you want to manage the organization roles for your users from within Grafana.
```

eleijonmarck commented on pkg/login/social/github_oauth.go:
```diff
@@ -201,14 +204,25 @@ func (s *SocialGithub) UserInfo(client *http.Client, token *oauth2.Token) (*Basi
+
+	if !s.skipOrgRoleSync {
+		var grafanaAdmin bool
+		role, grafanaAdmin = s.extractRoleAndAdmin(response.Body, teams, true)
```
comment: it is a necessity that we assign to role and grafanaAdmin here and not a new variable as the scope would "swallow" the value of the role in the scope.

eleijonmarck committed:
Update docs/sources/setup-grafana/configure-security/configure-authentication/github/index.md

eleijonmarck committed:
Update docs/sources/setup-grafana/configure-security/configure-authentication/github/index.md

eleijonmarck committed:
Merge branch 'main' into eleijonmarck/auth/github-skip-org-role-sync

eleijonmarck committed:
merge fix
